File: directcall-java/app/src/main/java/com/sendbird/calls/quickstart/utils/PushUtils.java
package com.sendbird.calls.quickstart.utils;

import android.content.Context;
import android.text.TextUtils;
import android.util.Log;

import com.google.firebase.iid.FirebaseInstanceId;
import com.sendbird.calls.SendBirdCall;
import com.sendbird.calls.SendBirdException;
import com.sendbird.calls.quickstart.BaseApplication;

public class PushUtils {

    public interface GetPushTokenHandler {
        void onResult(String token, SendBirdException e);
    }

    public static void getPushToken(Context context, final GetPushTokenHandler handler) {
        Log.i(BaseApplication.TAG, "[PushUtils] getPushToken()");

        String savedToken = PrefUtils.getPushToken(context);
        if (TextUtils.isEmpty(savedToken)) {
            FirebaseInstanceId.getInstance().getInstanceId().addOnCompleteListener(task -> {
                if (!task.isSuccessful()) {
                    Log.i(BaseApplication.TAG, "[PushUtils] getPushToken() => getInstanceId failed", task.getException());
                    if (handler != null) {
                        handler.onResult(null, new SendBirdException((task.getException() != null ? task.getException().getMessage() : "")));
                    }
                    return;
                }

                String pushToken = (task.getResult() != null ? task.getResult().getToken() : "");
                Log.i(BaseApplication.TAG, "[PushUtils] getPushToken() => pushToken: " + pushToken);
                if (handler != null) {
                    handler.onResult(pushToken, null);
                }
            });
        } else {
            Log.i(BaseApplication.TAG, "[PushUtils] savedToken: " + savedToken);
            if (handler != null) {
                handler.onResult(savedToken, null);
            }
        }
    }

    public interface PushTokenHandler {
        void onResult(SendBirdException e);
    }

    public static void registerPushToken(Context context, String pushToken, PushTokenHandler handler) {
        Log.i(BaseApplication.TAG, "[PushUtils] registerPushToken(pushToken: " + pushToken + ")");

        SendBirdCall.registerPushToken(pushToken, false, e -> {
            if (e != null) {
                Log.i(BaseApplication.TAG, "[PushUtils] registerPushToken() => e: " + e.getMessage());
                PrefUtils.setPushToken(context, pushToken);

                if (handler != null) {
                    handler.onResult(e);
                }
                return;
            }

            Log.i(BaseApplication.TAG, "[PushUtils] registerPushToken() => OK");
            PrefUtils.setPushToken(context, pushToken);

            if (handler != null) {
                handler.onResult(null);
            }
        });
    }
}

File: directcall-kotlin/app/src/main/java/com/sendbird/calls/quickstart/utils/PushUtils.kt
package com.sendbird.calls.quickstart.utils

import android.content.Context
import android.util.Log
import com.google.firebase.messaging.FirebaseMessaging
import com.sendbird.calls.SendBirdCall.registerPushToken
import com.sendbird.calls.SendBirdException
import com.sendbird.calls.internal.PushTokenType
import com.sendbird.calls.quickstart.TAG

fun Context.getPushToken(handler: GetPushTokenHandler?) {
    Log.i(TAG, "[PushUtils] getPushToken()")
    val savedToken = getPushToken()
    if (savedToken.isNullOrEmpty()) {
        FirebaseMessaging.getInstance().token.addOnCompleteListener { task ->
            val pushToken = task.result
            if (!task.isSuccessful || pushToken == null) {
                Log.i(TAG, "[PushUtils] getPushToken() => getInstanceId failed", task.exception)
                handler?.onResult(
                    null,
                    SendBirdException(if (task.exception != null) task.exception?.message else "")
                )
                return@addOnCompleteListener
            }
            Log.i(TAG, "[PushUtils] getPushToken() => pushToken: $pushToken")
            handler?.onResult(pushToken, null)
        }
    } else {
        Log.i(TAG, "[PushUtils] savedToken: $savedToken")
        handler?.onResult(savedToken, null)
    }
}

fun Context.registerPushToken(pushToken: String, handler: PushTokenHandler?) {
    Log.i(TAG, "[PushUtils] registerPushToken(pushToken: $pushToken)")
    registerPushToken(
        pushToken,
        PushTokenType.FCM_VOIP,
        false
    ) labelRegisterPushToken@{ e: SendBirdException? ->
        if (e != null) {
            Log.i(TAG, "[PushUtils] registerPushToken() => e: " + e.message)
            setPushToken(pushToken)
            handler?.onResult(e)
            return@labelRegisterPushToken
        }
        Log.i(TAG, "[PushUtils] registerPushToken() => OK")
        setPushToken(pushToken)
        handler?.onResult(null)
    }
}

fun interface GetPushTokenHandler {
    fun onResult(token: String?, e: SendBirdException?)
}

fun interface PushTokenHandler {
    fun onResult(e: SendBirdException?)
}